### PR TITLE
This will reduce "Already borrowed error":

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Original issue https://github.com/huggingface/tokenizers/issues/537

The original issue is caused by transformers calling many times
mutable functions on the rust tokenizers.
Rust needs to guarantee that only 1 agent has a mutable reference
to memory at a given time (for many reasons which don't need explaining
here). Usually, the rust compiler can guarantee that this property is
true at compile time.

Unfortunately, this is impossible for Python to do that, so PyO3, the
bridge between rust and python used by `tokenizers`, will change the
compile guarantee for a dynamic guarantee, so if multiple agents try
to have multiple mutable borrows at the same time, then the runtime will
yell with "Already borrowed".

The proposed fix here in transformers, is simply to reduce the actual
number of calls that really need mutable borrows. By reducing them,
we reduce the risk of running into "Already borrowed" error.
The caveat is now we add a call to read the current configuration of the
`_tokenizer`, so worst case we have 2 calls instead of 1, and best case
we simply have 1 + a Python comparison of a dict (should be negligible).

Fixes #https://github.com/huggingface/tokenizers/issues/537

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@n1t0 @LysandreJik

### CODE REVIEW

**Review by n1t0 — APPROVED** [review]
Nice, LGTM!

**Comment by Narsil** [comment]
@LysandreJik I would welcome your look on this too.

File: tests/test_tokenization_fast.py [hunk 1]
```diff
@@ -90,3 +91,21 @@ def test_training_new_tokenizer_with_special_tokens_change(self):
                     # is restored
                     shutil.rmtree(self.tmpdirname)
                     self.tmpdirname = tmpdirname_orig
+
+
+@require_tokenizers
+class ReduceMutableBorrowTests(unittest.TestCase):
+    def test_async_share_tokenizer(self):
+        tokenizer = PreTrainedTokenizerFast.from_pretrained("gpt2")
+        tokenizer.add_special_tokens({"pad_token": "[PAD]"})
```
Comment (SaulLu) [inline tests/test_tokenization_fast.py:100]: As I have the impression here that the test is not really dependent on the tokenizer tested, I suggest you use this dummy tokenizer which is much smaller than gpt2 and already has the token `[PAD]` filled in. :slightly_smiling_face: 
```suggestion
        tokenizer = PreTrainedTokenizerFast.from_pretrained("robot-test/dummy-tokenizer-wordlevel")
```

File: tests/test_tokenization_fast.py [hunk 2]
```diff
@@ -90,3 +91,21 @@ def test_training_new_tokenizer_with_special_tokens_change(self):
                     # is restored
                     shutil.rmtree(self.tmpdirname)
                     self.tmpdirname = tmpdirname_orig
+
+
+@require_tokenizers
+class ReduceMutableBorrowTests(unittest.TestCase):
+    def test_async_share_tokenizer(self):
```
Comment (SaulLu) [inline tests/test_tokenization_fast.py:98]: Great test! I could have been a very small suggestion: wouldn't we like to leave a comment to explain why we need to do this test? (maybe mentioning the issue #https://github.com/huggingface/tokenizers/issues/537)

**Review by SaulLu — COMMENTED** [review]
Thank you a lot for implementing this! :+1: 

I just left 2 micro comments on the test, tell me what you think!

**Review by LysandreJik — APPROVED** [review]
Great! Thanks a lot @Narsil!

File: tests/test_tokenization_fast.py (hunk 1 above)
Comment (Narsil) [reply tests/test_tokenization_fast.py:100]: Thanks !

**Pull request merged**